fix casting issue for upcoming relay-runtime types

mattkrick opened the pull request:
related to #89 

in the future, `@types/relay-runtime` may have stricter, more accurate typings.
if that happens, we should still spit out error-free ts.
this does that.

example of change
```typescript
const node: ReaderFragment = () => ({})() //old
const node = () => ({})() as ReaderFragment //new
````

this is necessary because:
- discriminated unions use string literals
- iifes coerce string literals to `string`s (eg `kind: "Request"` turns into `kind: string`)
- relay-compiler dedupes by sometimes compressing the document into an iife

alloy commented on src/formatGeneratedModule.ts:
```diff
@@ -18,8 +18,8 @@ export const formatterFactory = (
   const docTextComment = docText ? "\n/*\n" + docText.trim() + "\n*/\n" : "";
-  let nodeStatement = `const node: ${documentType ||
-    "never"} = ${concreteText};`;
+  let nodeStatement = `const node = ${concreteText} as ${documentType ||
```
Do you think it would make sense to wrap the `concreteText` in an extra set of parentheses? Just to ensure there can never be any ambiguity.

```suggestion
  let nodeStatement = `const node = (${concreteText}) as ${documentType ||
```

mattkrick replied:
i've got a soft preference to exclude unnecessary parens just because if someone lints their generated files with `no-extra-parens` then it'll have to rewrite the file every time. that said, it's easy enough excluding the folder from linting. just lemme know & i'll accept the change

alloy replied:
That _is_ a good point. However, I already think people are and should ignore these artefacts from their prettier/linter setup (we do), as there’s no way to make them fit everybody’s preferences. As such, I think we should optimise for building safe artefacts.

alloy commented:
Thanks! And sorry for the late reply, I must have done an over-zealous “mark all as read” of my notifications.

alloy commented:
I took the liberty to rebase your branch to get the tests to run on CI. There’s still [another issue](https://github.com/relay-tools/relay-compiler-language-typescript/pull/97) that eventually fails the build, but the tests now at least run and they point out that this change needs updated tests.

mattkrick commented:
thanks for the review! no worries on time, i still haven't PRed definitelytyped yet with the new types (they're pretty sweet though, finally got type checking proxies).

i'll fix up the tests as soon as we confirm the parens deal.

mattkrick commented:
fixed tests. no longer a 1-line PR but still not bad.

note: i removed `noImplicitAny`. if i understand that tension correctly, this solves that problem, too.

mattkrick commented on types/relay-compiler/language/RelayLanguagePluginInterface.d.ts:
```diff
@@ -22,12 +22,14 @@ export type FormatModule = (
       | typeof RelayConcreteNode.FRAGMENT
       | typeof RelayConcreteNode.REQUEST
       | typeof RelayConcreteNode.BATCH_REQUEST
+      | "ConcreteFragment"
```
ts was yelling at me about this one.
also future proofed stuff by adding in the 2 that are used today below

alloy replied:
Does this mean that in reality `typeof RelayConcreteNode.FRAGMENT` should be `"ConcreteFragment"`?

And just for my understanding, in what way was TS complaining about these?

mattkrick replied:
to be really accurate, the relay-compiler only creates 2 types now: ReaderFragment & ConcreteRequest. Batch requests got deprecated around v1.3 & tbh i don't know when fragment & request got deprecated.

mattkrick replied:
ts was complaining when it was typechecking the test file

alloy replied:
Ah, hmm, ok. Well, something to look into in the future, not in any way a blocker for this PR.

mattkrick commented on types/relay-compiler/language/RelayLanguagePluginInterface.d.ts:
```diff
@@ -22,12 +22,14 @@ export type FormatModule = (
     typeText: string;
-    hash: string | null;
-    devOnlyAssignments: string | null;
+    devOnlyAssignments?: string | null;
```
i saw a PR about this, so i left it in, but made it optional so ts won't yell at me

alloy replied:
According to the upstream interface, if my rusty Flow isn’t failing me, this property isn’t optional but does take `null` or `undefined` https://github.com/facebook/relay/blob/c0376a5bb06e0e1c3b62d662027b18c0322e84ec/packages/relay-compiler/language/RelayLanguagePluginInterface.js#L176

I’d like our TS version of the interface to accurately mimic the upstream version.

mattkrick replied:
ah, i didn't know there was a ref doc. should i put hash back in & add that to the tests? looks like there's some ambiguity with hash & sourceHash in the doc

alloy replied:
Saw you already did so, thanks!

alloy commented on test/formatGeneratedModule-test.ts:
```diff
@@ -37,28 +37,27 @@ expect.extend({
       concreteText: JSON.stringify({ the: { fragment: { data: 42 } }}),
       typeText: 'export type CompleteExample = { readonly id: string }',
-      hash: 'abcde',
+      relayRuntimeModule: 'relay-runtime',
```
This has been removed from the upstream plugin interface https://github.com/facebook/relay/commit/dce32eaa485238569c3060366b693c95b07d631f#diff-f5a244f9b04c4b73796c0bd3e1c43854L202, can you remove it from our TS version of the interface instead?

alloy commented:
> note: i removed `noImplicitAny`. if i understand that tension correctly, this solves that problem, too.

Awesome, great stuff 👌 

I do think it might be good for us to have a test  that our artefacts pass TS compilation with strict settings. Could you add a package script that would run `tsc --noEmit --pretty ./test/fixtures/generated-module/*.ts` and add that to the `test` script?

mattkrick commented:
> Could you expand the type-check package script to also check ./test/fixtures/generated-module/*.ts?

sorry, i don't follow. you want to use https://www.npmjs.com/package/type-check to compile the fixtures & see if they create typescript errors?

alloy commented:
No, sorry for the confusion, I meant https://github.com/relay-tools/relay-compiler-language-typescript/blob/bb0b6a165c7d556a265b20007aa651003274effe/package.json#L41

mattkrick commented:
whoa, let this one slip. @alloy all fixed up. there's still a handful of tests breaking on master, but this little bit should be good to go

maraisr commented on package.json:
```diff
@@ -29,7 +29,7 @@
       "https://github.com/relay-tools/relay-compiler-language-typescript/issues"
   },
-  "main": "lib/index.js",
+  "main": "lib/src/index.js",
```
Was this intentional?

mattkrick replied:
that's a great question! looks like an artifact from something i tried & forgot to revert.
sorry for letting this go stale, there's no errors with the latest relay-runtime types, so i'm happy to close

alloy replied:
Let’s do that then 👍
